python/pyarrow/_cuda.pyx

```diff
@@ -246,6 +283,73 @@ cdef class Context:
             result.copy_from_device(buf, position=0, nbytes=size)
         return result
 
+    def buffer_from_object(self, obj):
+        """Create device buffer view of arbitray device memory object.
+
+        When the object contains a non-contiguous view of device
+        memory then the returned device buffer will contain contiguous
+        view of the memory, that is, including the intermediate data
+        that is otherwise invisible to the input object.
+
+        Parameters
+        ----------
+        obj : object
+          Specify object that contains address to device memory or
+          implements cuda array interface.
+
+        Returns
+        -------
+        cbuf : CudaBuffer
+          Device buffer as a view of device memory.
+
+        """
+        import numpy as np
+        if isinstance(obj, HostBuffer):
+            if not obj.address and obj.size == 0:
+                return self.new_buffer(0)
+            return self.foreign_buffer(obj.address, obj.size)
+        if isinstance(obj, Buffer):
+            obj = CudaBuffer.from_buffer(obj)
+        if isinstance(obj, CudaBuffer):
+            if not obj.address and obj.size == 0:
+                return self.new_buffer(0)
+            return self.foreign_buffer(obj.address, obj.size)
+        elif hasattr(obj, '__cuda_array_interface__'):
+            desc = obj.__cuda_array_interface__
+            shape = desc['shape']
+            strides = desc.get('strides')
+            dtype = np.dtype(desc['typestr'])
+            addr = desc['data'][0]
+            if addr is None:
+                nbytes = 0
+            elif not strides:
+                nbytes = dtype.itemsize * product(shape)
+            else:
+                start = 0
+                end = dtype.itemsize
+                for i, dim in enumerate(shape):
+                    if dim == 0:
+                        start = end = addr
+                        break
+                    stride = strides[i]
+                    if stride > 0:
+                        end += stride * (dim - 1)
+                    elif stride < 0:
+                        start += stride * (dim - 1)
+                nbytes = end - start
+            if nbytes:
+                return self.foreign_buffer(addr + start, nbytes)
+            else:
+                return self.new_buffer(0)
```

Comment:
So... should `foreign_buffer(0, 0)` be supported instead?

Reply:
Done.